src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1402,23 +1573,362 @@ DatabaseShardImp::storeLedgerInShard(
+                    {
+                        moveShard(*cur);
+                    }
+                }
+
+                cur = boost::none;
+            }
+        }
+    }
+}
+
+auto
+DatabaseShardImp::prepareForNewShard(
+    std::uint32_t shardIndex,
+    std::uint32_t numHistoricalShards,
+    std::lock_guard<std::mutex> const& lock) -> boost::optional<PathDesignation>
+{
+    // Any shard earlier than the two most recent shards
+    // is a historical shard
+    auto const boundaryIndex = shardBoundaryIndex(lock);
+    auto const isHistoricalShard = shardIndex < boundaryIndex;
+
+    auto const designation = isHistoricalShard && !historicalPaths_.empty()
+        ? PathDesignation::historical
+        : PathDesignation::none;
+
+    // Check shard count and available storage space
+    if (isHistoricalShard && numHistoricalShards >= maxHistoricalShards_)
+    {
+        JLOG(j_.error()) << "maximum number of historical shards reached";
+        canAdd_ = false;
+        return boost::none;
+    }
+    if (!sufficientStorage(1, designation, lock))
+    {
+        JLOG(j_.error()) << "insufficient storage space available";
+        canAdd_ = false;
+        return boost::none;
+    }
+
+    return designation;
+}
+
+boost::filesystem::path
+DatabaseShardImp::chooseHistoricalPath(std::lock_guard<std::mutex> const&) const
+{
+    // If not configured with separate historical paths,
+    // use the main path (dir_) by default.
+    if (historicalPaths_.empty())
+        return dir_;
+
+    boost::filesystem::path historicalShardPath;
+    std::vector<boost::filesystem::path> potentialPaths;
+
+    for (boost::filesystem::path const& path : historicalPaths_)
+    {
+        if (boost::filesystem::space(path).available >= avgShardFileSz_)
+            potentialPaths.push_back(path);
+    }
+
+    if (potentialPaths.empty())
+    {
+        JLOG(j_.error()) << "failed to select a historical shard path";
+        return "";
+    }
+
+    std::sample(
+        potentialPaths.begin(),
+        potentialPaths.end(),
+        &historicalShardPath,
+        1,
+        default_prng());
+
+    return historicalShardPath;
+}
+
+bool
+DatabaseShardImp::checkHistoricalPaths() const
+{
+#if BOOST_OS_LINUX
+    // Each historical shard path must correspond
+    // to a directory on a distinct device or filesystem.
+    // Currently, this constraint is enforced only on
+    // Linux.
+
+    std::unordered_set<int> filesystemIDs;
+    for (auto const& path : historicalPaths_)
+    {
+        struct statvfs buffer;
+        if (statvfs(path.c_str(), &buffer))
+        {
+            JLOG(j_.error())
+                << "failed to acquire stats for a 'historical_shard_path'";
+            return false;
+        }
+
+        if (!filesystemIDs.insert(buffer.f_fsid).second)
+        {
+            JLOG(j_.error())
+                << "each 'historical_shard_path' must correspond to a "
+                   "distinct device or filesystem";
```

Comment:
I think it would be worthwhile to output the duplicate path here, too. But now that I look at it, it would be even better to output both paths that are on the same device so the user can pick one. I think the best way to implement this would be to use a `map` similar to how you checked below for non-linux filesystems.

Reply:
Done